core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapYearInComparison.java

```diff
@@ -0,0 +1,232 @@
+                // Cannot replace with a range due to how year operates on value's local date/time.
+                // I.e. unwrapping is possible only when values are all of some fixed zone and the zone is known.
+                return expression;
+            }
+
+            return switch (expression.getOperator()) {
+                case EQUAL -> between(argument, argumentType, calculateRangeStartInclusive(right, argumentType), calculateRangeEndInclusive(right, argumentType));
+                case NOT_EQUAL -> new NotExpression(between(argument, argumentType, calculateRangeStartInclusive(right, argumentType), calculateRangeEndInclusive(right, argumentType)));
+                case IS_DISTINCT_FROM -> or(
+                        new IsNullPredicate(argument),
+                        new NotExpression(between(argument, argumentType, calculateRangeStartInclusive(right, argumentType), calculateRangeEndInclusive(right, argumentType))));
+                case LESS_THAN -> new ComparisonExpression(LESS_THAN, argument, toExpression(calculateRangeStartInclusive(right, argumentType), argumentType));
+                case LESS_THAN_OR_EQUAL -> new ComparisonExpression(LESS_THAN_OR_EQUAL, argument, toExpression(calculateRangeEndInclusive(right, argumentType), argumentType));
+                case GREATER_THAN -> new ComparisonExpression(GREATER_THAN, argument, toExpression(calculateRangeEndInclusive(right, argumentType), argumentType));
+                case GREATER_THAN_OR_EQUAL -> new ComparisonExpression(GREATER_THAN_OR_EQUAL, argument, toExpression(calculateRangeStartInclusive(right, argumentType), argumentType));
+            };
+        }
+
+        private Object calculateRangeStartInclusive(Object yearVariable, Type type)
+        {
+            int year = toIntExact((Long) yearVariable);
+            if (type == DATE) {
+                LocalDate firstDay = LocalDate.ofYearDay(year, 1).with(firstDayOfYear());
+                return firstDay.toEpochDay();
+            }
+            if (type instanceof TimestampType timestampType) {
+                if (timestampType.isShort()) {
+                    LocalDateTime dateTime = LocalDateTime.MIN.withYear(year).with(firstDayOfYear());
+                    return dateTime.toEpochSecond(ZoneOffset.UTC) * MICROSECONDS_PER_SECOND + LongMath.divide(dateTime.getNano(), NANOSECONDS_PER_MICROSECOND, UNNECESSARY);
+                }
+                LocalDateTime dateTime = LocalDateTime.MIN.withYear(year).with(firstDayOfYear());
+                long endInclusiveMicros = dateTime.toEpochSecond(ZoneOffset.UTC) * MICROSECONDS_PER_SECOND
+                        + LongMath.divide(dateTime.getNano(), NANOSECONDS_PER_MICROSECOND, UNNECESSARY);
+                return new LongTimestamp(endInclusiveMicros, 0);
+            }
+            throw new UnsupportedOperationException("Unsupported type: " + type);
+        }
+
+        private Object calculateRangeEndInclusive(Object yearVariable, Type type)
```

Comment:
I would keep as-is because the argument is different. `UnwrapDateTruncInComparison` takes epoch day/micros and `UnwrapYearInComparison` takes year.

Reply:
Among `rangeUnit` there is also `YEAR`. My intention was to avoid duplicating the rather complex logic of dealing with calculating the inclusive end range value.